CP-33044 implement NVML attach/detach

lindig opened the pull request:
The NVML library is required to query Nvidia GPUs at runtime. This library is loaded dynamically using dlopen(2). Add the capability to attach and detach this library using the xapi-idl IPC mechanism. This is experimental and replaced the existing mock implementation.

lindig commented:
The new RPC calls are not yet used by clients but can be exercised using `gpumon-cli`, which is part of xen-api.git.

lindig commented:
This has passed a BST and I have tested manually the attach/detach feature - it can be exercised using `gpumon-cli`, which is now part of xen-api.git.

psafont commented on lib/nvml.ml:
```diff
@@ -136,13 +138,59 @@ let get_vgpu_for_uuid iface vgpu_uuid vgpus =
+    let interface = library_open () in
+    try init interface ; interface with e -> library_close interface ; raise e
+
+  let get () = !interface
```
I wonder if it would be wiser to have

```ocaml
val with_nvml : (interface option -> unit) -> unit
```
That can holds and drops the mutex related to the interface handle to avoid race conditions

lindig replied:
Currently the mutex is only held while updating the value that holds the interface. Do you suggest to hold it while the interface/library is loaded?

psafont replied:
I meant holding the mutex when the interface is being read, would that cause problems because there's more than one thread trying to read it?

psafont commented on gpumon/gpumon.ml:
```diff
@@ -33,14 +33,17 @@ let acquire_nvml_interface ?(timeout = 30.0) () =
+    | None ->
+        Process.D.info "Failed to open NVML interface - retrying in %4.0f"
+          delay ;
+        Thread.delay delay ;
+        try_again_after (min (delay *. 1.5) (20.0 *. 60.0))
```
Should this wait be done _after_ releasing the interface instead?

psafont replied:
I misunderstood, don't worry about the commet

robhoes commented on lib/dune:
```diff
@@ -1,4 +1,4 @@
  (name gpumon_lib)
  (wrapped false)
- (libraries nvml_stubs rresult xapi-stdext-unix))
+ (libraries nvml_stubs xapi-idl rresult xapi-stdext-unix))
```
If this is just for the Debug module, then you can use xapi-log instead.

robhoes commented on lib/nvml.ml:
```diff
@@ -136,13 +138,59 @@ let get_vgpu_for_uuid iface vgpu_uuid vgpus =
+    with_mutex mx @@ fun () ->
+    match !interface with
+    | Some _ ->
+        D.warn "Nvml library attach: library already attached"
```
Attach and detach are idempotent, which avoids trouble with interference between the RRD thread and the RCP calls.

robhoes commented on gpumon/gpumon.ml:
```diff
@@ -33,14 +33,17 @@ let acquire_nvml_interface ?(timeout = 30.0) () =
 (* acquire_nvml_interface but keep looping. Only call this in a thread
    that does not block the main interaction like the RPC server *)
 let rec always_acquire_nvml_interface () =
```
Could drop `rec` here now that the inner function has the recursive calls.

lindig commented:
It would be safe to merge this as this API is currently not exercised from the xapi side. It requires a small change in xapi, but it needs to be tested. Manual testing demonstrates that this code works. 

```
diff --git a/ocaml/xapi/xapi_gpumon.ml b/ocaml/xapi/xapi_gpumon.ml
index 384ab68f8..4ed7bb269 100644
--- a/ocaml/xapi/xapi_gpumon.ml
+++ b/ocaml/xapi/xapi_gpumon.ml
@@ -18,29 +18,20 @@ open D
 
 let gpumon = "xcp-rrdd-gpumon"
 
-module Gpumon = Daemon_manager.Make (struct
-  let check =
-    Daemon_manager.Function
-      (fun () ->
-        try
-          ignore
-            (Forkhelpers.execute_command_get_output !Xapi_globs.systemctl
-               ["is-active"; "-q"; gpumon]
-            ) ;
-          true
-        with _ -> false
-      )
-
-  let start () =
-    debug "Starting %s" gpumon ;
-    Xapi_systemctl.start ~wait_until_success:false gpumon
-
-  let stop () =
-    debug "Stopping %s" gpumon ;
-    Xapi_systemctl.stop ~wait_until_success:false gpumon
-end)
-
-let with_gpumon_stopped = Gpumon.with_daemon_stopped
+(** we no longer stop gpumon but simply tell it to unload the NVML
+    library, which was the reason we needed to stop it *)
+let with_gpumon_stopped ?(timeout = 30.0) f =
+  let module GPU = Gpumon_client.Client.Nvidia in
+  let defer finally = Fun.protect ~finally in
+  let dbg = __FUNCTION__ in
+  match GPU.nvml_is_attached dbg with
+  | false ->
+      (* nothing to do, just execute f *)
+      f ()
+  | true ->
+      (* detach, execute f, re-attach in any case *)
+      defer (fun () -> GPU.nvml_attach dbg) @@ fun () ->
+      GPU.nvml_detach dbg ; f ()
 
 module Nvidia = struct
   let key = "nvidia"
```

lindig commented:
I'm running some tests - want them to finish before I merge this.

lindig commented:
I have added some clarifying comments and changed some names: the RRD thread never attaches the NVML library. Outside from startup, this is only done via the RPC attach/detach call controlled by xapi or the CLI. So we never run into the situation that the library is re-attached behind xapi's back.

lindig commented:
This passed 178257 (Dev Run) but we can hold it back and merge the gpumon changes first and independently.

lindig commented:
This is still undergoing testing. I've seen some failures to attach the NVML library coming from the C layer and I don't yet understand the circumstances. It would be still to merge this since we don't use the API but obviously if would be better to understand this.

lindig commented:
This is safe to merge. It is true that some testing sees failures but the reason is not related to the implementation: when a GPU is not bound to the Nvidia driver in the kernel (because it is passed through), NVML attach will fail. Maybe this could be better detected.

The latest changes make sure a failure to attach the library is properly propagated back to the API client, which previously it was not. The error from the underlying NVML C library in the case described above is a rather unspecific "unknown Error".

It is not clear what happens when a host has 2 GPUs, one in pass-through and one bound to the NVidia driver.
